lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -5866,6 +5883,50 @@ void TypeChecker::checkConformancesInContext(DeclContext *dc,
     if (tracker)
       tracker->addUsedMember({conformance->getProtocol(), Identifier()},
                              defaultAccessibility > Accessibility::FilePrivate);
+
+    // Diagnose @NSCoding on file/fileprivate/nested/generic classes, which
+    // have unstable archival names.
+    if (auto classDecl = dc->getAsClassOrClassExtensionContext()) {
+      if (isNSCoding(conformance->getProtocol())) {
+        // Note: these 'kind' values are synchronized with
+        // diag::nscoding_unstable_mangled_name.
+        Optional<unsigned> kind;
+        if (classDecl->getGenericSignature()) {
+          kind = 3;
+        } else if (classDecl->getDeclContext()->isTypeContext()) {
+          kind = 2;
+        } else {
+          switch (classDecl->getFormalAccess()) {
```

Comment:
That seems... awkward. `AccessScope` is great when we're trying to determine accessibility of some entity from some specific context. We're asking a simpler question about the stated access level.